dali/pipeline/operators/decoder/nvjpeg_decoder_new.h

```diff
@@ -0,0 +1,267 @@
+#define DALI_PIPELINE_OPERATORS_DECODER_NVJPEG_DECODER_NEW_H_
+
+#include "dali/pipeline/operators/operator.h"
+#include "dali/pipeline/operators/decoder/nvjpeg_decoupled.h"
+#include "dali/pipeline/operators/decoder/nvjpeg_helper.h"
+#include "dali/util/image.h"
+#include "dali/util/ocv.h"
+#include "dali/image/image_factory.h"
+
+namespace dali {
+
+class nvJPEGDecoderNew : public Operator<MixedBackend> {
+ public:
+  explicit nvJPEGDecoderNew(const OpSpec& spec) :
+    Operator<MixedBackend>(spec),
+    output_image_type_(spec.GetArgument<DALIImageType>("output_type")),
+    output_shape_(batch_size_),
+    output_info_(batch_size_),
+    image_decoders_(batch_size_),
+    image_host_states_(batch_size_),
+    jpeg_streams_(batch_size_),
+    pinned_buffer_(batch_size_),
+    decoder_host_state_h_(batch_size_),
+    decoder_huff_hybrid_state_h_(batch_size_) {
+    
+    NVJPEG_CALL(nvjpegCreateSimple(&handle_));
+
+    size_t device_memory_padding = spec.GetArgument<Index>("device_memory_padding");
+    size_t host_memory_padding = spec.GetArgument<Index>("host_memory_padding");
+    NVJPEG_CALL(nvjpegSetDeviceMemoryPadding(device_memory_padding, handle_));
+    NVJPEG_CALL(nvjpegSetPinnedMemoryPadding(host_memory_padding, handle_));
+
+    // to create also in GPU Op
+    NVJPEG_CALL(nvjpegDecoderCreate(
+                      handle_, NVJPEG_BACKEND_HYBRID, &decoder_huff_host_));
+    NVJPEG_CALL(nvjpegDecoderCreate(
+                      handle_, NVJPEG_BACKEND_GPU_HYBRID, &decoder_huff_hybrid_));  // NVJPEG_BACKEND_GPU_HYBRID
+
+  
+    for (int i = 0; i < batch_size_; i++) {
+      NVJPEG_CALL(nvjpegDecodeParamsCreate(handle_, &decode_params_[i]));
+      NVJPEG_CALL(nvjpegnvjpegDecodeParamsSetOutputFormat(decode_params_[i],
+                                            GetFormat(output_image_type_)));
+
+      NVJPEG_CALL(nvjpegJpegStreamCreate(handle_, &jpeg_streams_[i]));
+      // We want to use nvJPEG default pinned allocator
+      NVJPEG_CALL(nvjpegBufferPinnedCreate(handle_, nullptr, &pinned_buffer_[i]));
+
+      NVJPEG_CALL(nvjpegDecoderCreateStateHost(handle_, decoder_huff_host_, &decoder_host_state_h_[i]));
+      NVJPEG_CALL(nvjpegStateAttachPinnedBuffer(decoder_host_state_h_[i], pinned_buffer_[i]));
+
+      NVJPEG_CALL(nvjpegDecoderCreateStateHost(
+                               handle_, decoder_huff_hybrid_, &decoder_huff_hybrid_state_h_[i]));
+      NVJPEG_CALL(nvjpegStateAttachPinnedBuffer(
+                               decoder_huff_hybrid_state_h_[i], pinned_buffer_[i]));
+    }
+
+    // GPU
+    // create the handles, streams and events we'll use
+    // We want to use nvJPEG default device allocator
+    NVJPEG_CALL(nvjpegBufferDeviceCreate(handle_, nullptr, &device_buffer_));
+    NVJPEG_CALL(nvjpegDecoderCreateStateDevice(handle_, decoder_huff_host_, &decoder_state_d_));
+    NVJPEG_CALL(nvjpegStateAttachDeviceBuffer(decoder_state_d_, device_buffer_));
+  }
+
+  ~nvJPEGDecoderNew() noexcept(false) {
```

Comment:
virtual, we will probably want to specialize for the different cache types

Reply:
Done